### PR TITLE
[WIP] Cura 8640 PyQt6 upgrade

### PR DESCRIPTION
**Pull request by jellespijker**
In order to get this to work on our build-system and working for all three OSes we did a shit tons of boy scouting in our cmake. We removed old methods with variables and try to be consisted in a target-based approach. The idea is that we don't patch stuff down the line, but that the install should place everything with desired rpaths for each project.

Most of the changes had to do with how dependency targets were named and to make sure that external projects weren't downloaded automatically.

# Part of
- ultimaker/libarcus#134
- ultimaker/pynest2d#17
- ultimaker/libsavitar#41
- ultimaker/libnest2d#3
- ultimaker/cura-binary-data#17
- ultimaker/libcharon#40
- ultimaker/curaengine#1639
- ultimaker/uranium#804
- ultimaker/cura#11792
- ultimaker/cura-build#280
- ultimaker/cura-rundeck-jobs#1

# Fixes
- #79 We now use the System based Python
- #122 We use 3.10.4 now as base
- #114 No longer build in Docker on Windows
- #110 No longer build in Docker on Windows
- #111 We no longer use Jenkins
- #122 
- #97 Yeah
- #95 
- #90 Compiling with MSVC Visual Studio 2019 and 2022 works, WSL and/or Docker on Windows are no longer supported
- #89 Docker on Windows no longer supported. We are currently building in a Windows VM using the x64 MSVC Visual Studio 2019 or 2022
- #81 
- #78 

# Todo
- [x] Update Readme. But maybe best done in a separate ticket to prioritize with our other work for the 5.0 release. See #121 and CURA-9047 
- [x] Make sure the additional VC redist files are installed as described in ultimaker/cura-binary-data#17
- [x] Check and fix if confirmed CURA-9105 Uninstall Cura.exe doesn't uninstall the program
- [x] Check and fix if confirmed CURA-9103 Order of files in installation folder scrambled
- [x] Clean up and remove setting the branches and versioning in the conanfile.py
- [x] Remove Windows Docker files since we now build directly on a VM with Visual Studio 2019 and/or Visual Studio 2020 compiler.
- [x] Remove old environment scripts
- [x] Remove Jenkinsfile 
- [x] Create a cmake target which allows the installation of a virtual environment (excluding Cura, Uranium and CuraEngine)
- [x] Create installer for MSI. Might be best to do this in a seperate ticket after the 5.0 release

### CODE REVIEW

File: docker/linux/clang/Dockerfile [hunk 1]
```diff
@@ -0,0 +1,159 @@
...
+    && apt-get -qq update \
+    && apt-get -qq install -y --no-install-recommends --no-install-suggests -o=Dpkg::Use-Pty=0 clang-10 lld-10 libc++-10-dev \
+    && wget -q --no-check-certificate https://github.com/llvm/llvm-project/archive/llvmorg-${CLANG_VERSION}.tar.gz \
+    && tar zxf llvmorg-${CLANG_VERSION}.tar.gz \
+    && sed -E -i 's/OUTPUT_NAME\s+"unwind"/OUTPUT_NAME "llvm-unwind"/g' llvm-project-llvmorg-${CLANG_VERSION}/libunwind/src/CMakeLists.txt \
+    && sed -i 's/unwind/llvm-unwind/g' llvm-project-llvmorg-${CLANG_VERSION}/clang/lib/Driver/ToolChains/CommonArgs.cpp \
+    && cd llvm-project-llvmorg-${CLANG_VERSION} \
+    && mkdir build \
+    && cd build/ \
+    && cmake ../llvm \
```
Comment (Ghostkeeper) [inline docker/linux/clang/Dockerfile:39]: Why are we compiling LLVM here? Is `apt get clang` not sufficient? It would make this docker file a lot simpler, especially with having to mess with those temporary files below.

File: README.md [hunk 2]
```diff
@@ -1,283 +1,221 @@
...
-      -DCMAKE_INSTALL_PREFIX=$cbe_install_dir \
-      -DCMAKE_PREFIX_PATH=$cbe_install_dir \
-      ..
-make
-```
-
-Note that a fairly recent C++ compiler is required, at the very least GCC
-4.9. If you are building on CentOS, you can get this from the devtoolset
-packages.
+This repository also contains dockerfiles which will build a cura-builder image for either Windows or Linux. These images
```
Comment (Ghostkeeper) [inline README.md:163]: Windows has been removed.
```suggestion
This repository also contains dockerfiles which will build a cura-builder image for Linux. These images
```

File: README.md [hunk 3]
```diff
@@ -1,283 +1,221 @@
...
+contain all build essentials to both build cura-build-environment and cura-build
```
Comment (Ghostkeeper) [inline README.md:164]:
```suggestion
contain all build essentials to build Cura.
```

File: projects/cura.cmake [hunk 4]
```diff
@@ -0,0 +1,75 @@
...
+GetFromEnvironmentOrCache(
+        NAME
+            CURA_CLOUD_ACCOUNT_API_ROOT
+        DESCRIPTION
+            "The cloud account API root")
+GetFromEnvironmentOrCache(
+        NAME
+            CURA_DIGITAL_FACTORY_URL
+        DESCRIPTION
+            "The Digit factory url")
```
Comment (Ghostkeeper) [inline projects/cura.cmake:31]:
```suggestion
            "The Ultimaker Digital Factory URL")
```

File: projects/cura.cmake [hunk 5]
```diff
@@ -0,0 +1,75 @@
...
+GetFromEnvironmentOrCache(
+        NAME
+            CURA_MARKETPLACE_ROOT
+        DESCRIPTION
+            "The marketplace root")
```
Comment (Ghostkeeper) [inline projects/cura.cmake:36]:
```suggestion
            "The Cura Marketplace API root URL")
```

File: projects/curaengine.cmake [hunk 6]
```diff
@@ -0,0 +1,30 @@
+# Copyright (c) 2022 Ultimaker B.V.
+# cura-build is released under the terms of the AGPLv3 or higher.
+
+GetFromEnvironmentOrCache(
+		NAME
```
Comment (Ghostkeeper) [inline projects/curaengine.cmake:5]: This file is a mix of tab-indented and space-indented. For our code style we have to indent everything with 4 spaces per level.

**Review by Ghostkeeper — CHANGES_REQUESTED** [review]

File: CMakeLists.txt [hunk 7]
```diff
@@ -1,74 +1,105 @@
...
 include(ExternalProject)
 include(GNUInstallDirs)
+include(ProcessorCount)
 
-include(BuildPlatformDetection)
-include(SetProjectDependencies)
+ProcessorCount(N_PROC)
+if(N_PROC EQUAL 0)
+    set(N_PROC 1)
+endif()
```
[the remark this answers is not in the document]
Comment (jellespijker) [reply CMakeLists.txt:27]: It is better to use Ninja as a generator, which ensured that all make jobs used the maximum number of threads, but some people also use make which uses 1 as a default.

I think this is a remnant of some old autobuild jobs we had before. All our current externals are CMake or Conan. So in that sense we can remove this. But I will double check that first.

File: CMakeLists.txt [hunk 8]
```diff
@@ -1,74 +1,105 @@
...
+        DESCRIPTION
+            "Cura pre release tag Version Information a, b or rc")
+GetFromEnvironmentOrCache(
+        NAME
+            CURA_VERSION_BUILD
+        DESCRIPTION
+            "Cura Version build")
+set(CURA_VERSION_EXTRA )
+if(NOT ${CURA_VERSION_PRE_RELEASE_TAG} STREQUAL "")
+    set(CURA_VERSION_EXTRA "-${CURA_VERSION_PRE_RELEASE_TAG}+${CURA_VERSION_BUILD}")
```
[the remark this answers is not in the document]
Comment (jellespijker) [reply CMakeLists.txt:63]: The build metadata tag in SemVer is most commonly a increasing version number. But after reading https://semver.org/ 
> Build metadata MAY be denoted by appending a plus sign and a series of dot separated identifiers immediately following the patch or pre-release version. Identifiers MUST comprise only ASCII alphanumerics and hyphens [0-9A-Za-z-]. Identifiers MUST NOT be empty. Build metadata MUST be ignored when determining version precedence. Thus two versions that differ only in the build metadata, have the same precedence. Examples: 1.0.0-alpha+001, 1.0.0+20130313144700, 1.0.0-beta+exp.sha.5114f85, 1.0.0+21AF26D3—-117B344092BD.

And since we now have to manually increase this number in a rundeck job I think a timestamp might be a good intermittent solution before our switch to Conan and Pipelines.

Special care should be taken to make sure that timestamp is used for all three OSes. that requires us to set it in Rundeck and not here.

@evtrados Can you make sure that all three build jobs on Rundeck use the same timestamp ad the CURA_VERSION_BUILD

File: projects/curaengine.cmake (hunk 6 above)
Comment (jellespijker) [reply projects/curaengine.cmake:5]: Yeah somehow my IDE did some weirdness in a few files. Will look into it

File: README.md [hunk 9]
```diff
@@ -1,283 +1,221 @@
...
+```
+> NOTE:
+> By specifying the conan data folder we ensure that big dependencies, such as Python
+> and Boost can be used in multiple environments and are reused when we are recreating
+> the environment from scratch.
+
+We can then build the environment with:
+```bash
+cd <root_cura_build_environment>
+docker run cura-env-builder:cura-8640 \
```
[the remark this answers is not in the document]
Comment (jellespijker) [reply README.md:197]: README.md needs a big revision this is no longer used as such but we already have CURA-9047 for this

File: README.md [hunk 10]
```diff
@@ -1,283 +1,221 @@
...
-      -G "NMake Makefiles" ^
-      ..
-nmake
+```bash
+docker create -t cura-builder:<branch_name \
+-v <source_path_cura-build-environment>:/home/ultimaker/source \
+-v <build_path_cura-build-environment>:/home/ultimaker/build \
+-v <install_path_cura-build-environment>:/home/ultimaker/install \
+-v <install_path_cura-build-environment>:/home/ultimaker/env \
+-v ~/.conan/data:/home/ultimaker/.conan/data cura-build-env:<branch_name
```
[the remark this answers is not in the document]
Comment (jellespijker) [reply README.md:211]: see above

File: cmake/nsis.cmake [hunk 11]
```diff
@@ -0,0 +1,38 @@
...
+            ${CMAKE_SOURCE_DIR}/packaging/cura_license.txt
+            LZMA
+            ${CMAKE_SOURCE_DIR}/packaging/cura_banner_nsis.bmp
+            ${CMAKE_SOURCE_DIR}/packaging/Cura.ico
+            ${installer_DIR}/dist/${INSTALLER_FILENAME}
+        COMMAND
+            makensis /V2 /P4 ${installer_DIR}/dist/Ultimaker-Cura.nsi
+        COMMENT  "Package into a NSIS installer."
+        )
+#add_dependencies(packaging pyinstaller)
```
[the remark this answers is not in the document]
Comment (jellespijker) [reply cmake/nsis.cmake:38]: yep, this is a left over debugging statement. Will uncomment

File: docker/linux/.env [hunk 12]
```diff
@@ -0,0 +1,29 @@
...
+CMAKE_VERSION=3.23
+PYTHON_VERSION=3.10
+PYTHON_VERSION_FULL=3.10.4
+LIBSTDCPP_VERSION=10.3.0
+LIBSTDCPP_PATCH_VERSION=28
+LIBSTDCPP_MAJOR_VERSION=10
+
+# Exact compiler versions
+GCC5_VERSION=5.5.0
+GCC6_VERSION=6.5.0
```
[the remark this answers is not in the document]
Comment (jellespijker) [reply docker/linux/.env:16]: Repurposed this docker-compose
https://github.com/conan-io/conan-docker-tools/tree/master/modern

File: docker/linux/base/Dockerfile [hunk 13]
```diff
@@ -0,0 +1,182 @@
...
+       libwayland-client0 \
+       libwayland-egl1 \
+    && rm -rf /var/lib/apt/lists/* \
+    && useradd -ms /bin/bash conan \
+    && printf "conan:conan" | chpasswd \
+    && adduser conan sudo \
+    && groupadd ultimaker \
+    && usermod -a -G ultimaker conan \
+    && printf "conan ALL= NOPASSWD: ALL\\n" >> /etc/sudoers \
+    && wget -q --no-check-certificate https://github.com/Kitware/CMake/releases/download/v${CMAKE_VERSION_FULL}/cmake-${CMAKE_VERSION_FULL}.tar.gz \
```
[the remark this answers is not in the document]
Comment (jellespijker) [reply docker/linux/base/Dockerfile:126]: lowers the complexity

**Comment by jellespijker** [comment]
Thanks @Ghostkeeper 

TLDRl; I'm reluctant to implement your bigger changes. Certainly given the time constraints we are already facing. I will change the smaller comments and I think @evtrados should implement the build version number on the Rundeck side

Yes you're correct with regards in the Dockerfile. These are based on the Conan Docker tools and could be simplified for our purpose. Something which I would certainly do if this repo still has a long life in front of it. But because we're switching to a new build setup with Conan as dependency manager and pipelines to build the dependencies and store them on an artifact server. All artifacts will be build on VM's instead of using a Docker Image. 

The same goes with the readme We should probably update this, certainly after the beta release. But it will be very minimal.
Basically stating that CMake needs to be called with: `CMAKE_PREFIX_PATH` and `CMAKE_INSTALL_PREFIX` since that will be basically what we do on the build servers as well. Okay I will also state the minimum requirements: CMake, Conan, Python, GCC, Apple-Clang or MSVC but no elaborate docs.

I see value in the build version remark for our current build system, but I do think that this should be handled on the Rundeck side, such that all three OS build jobs will get the same timestamp as build version. I'm hoping @evtrados can pick this up

File: packaging/NSIS/nsis-configurator.py [hunk 14]
```diff
@@ -0,0 +1,51 @@
...
+        main_app = sys.argv[4],
+        version_major = sys.argv[5],
+        version_minor = sys.argv[6],
+        version_patch = sys.argv[7],
+        version_build = sys.argv[8],
+        company = sys.argv[9],
+        web_site = sys.argv[10],
+        year = datetime.now().year,
+        cura_license_file = Path(sys.argv[11]),
+        compression_method = sys.argv[12],  # ZLIB, BZIP2 or LZMA
```
[the remark this answers is not in the document]
Comment (jellespijker) [reply packaging/NSIS/nsis-configurator.py:56]: different speeds and compression rates. We could switch between ZLIB (for our own internal builds) and LZMA  for distribution

File: docker/linux/gcc/Dockerfile [hunk 15]
```diff
@@ -0,0 +1,70 @@
...
+       dh-autoreconf \
+       libffi-dev \
+       libssl-dev \
+       pkg-config
+
+RUN wget -q --no-check-certificate http://mirrors.concertpass.com/gcc/releases/gcc-${GCC_VERSION}/gcc-${GCC_VERSION}.tar.xz \
+    && tar Jxf gcc-${GCC_VERSION}.tar.xz
+
+RUN cd gcc-${GCC_VERSION} \
+    && ./configure --build=x86_64-linux-gnu \
```
[the remark this answers is not in the document]
Comment (jellespijker) [reply docker/linux/gcc/Dockerfile:37]: To ensure we get the one we specify in the env file.

File: packaging/NSIS/nsis-configurator.py [hunk 16]
```diff
@@ -0,0 +1,51 @@
+import shutil
+import sys
+from datetime import datetime
+
+from pathlib import Path
+
+from jinja2 import Template
+
+if __name__ == "__main__":
```
[the remark this answers is not in the document]
Comment (Ghostkeeper) [reply packaging/NSIS/nsis-configurator.py:9]: That's why I wrote it like this ;) But maybe resolve/confirm the places where I put a question mark.

**Comment by jellespijker** [comment]
Created a ticket CURA-9157 for the MSI installer

**Comment by jellespijker** [comment]
Removing the CMake variables for the branches is best done by @evtrados when he creates the Rundeck scripts

**Comment by jellespijker** [comment]
See CURA-9105 for the uninstaller

**Comment by jellespijker** [comment]
I have updated the readme with the minimal information.

I have left out the Docker use case Since that is basically just a tool for us, to have a reproducible environment.

**Review by Ghostkeeper — APPROVED** [review]
